### ISSUE TITLE
TASK [ceph-osd : wait for all osd to be up]

### ISSUE THREAD

**Issue by ds2d** [ceph/ceph-ansible#5400]
Ubuntu 18.04.3
ceph_stable_release：nautilus
* master

fatal: [node3]: FAILED! => 
  msg: 'The conditional check ''(wait_for_all_osds_up.stdout | from_json)["osdmap"]["num_osds"] | int > 0'' failed. The error was: error while evaluating conditional ((wait_for_all_osds_up.stdout | from_json)["osdmap"]["num_osds"] | int > 0): ''dict object'' has no attribute ''num_osds'''

**Comment by dsavineau** [comment]
Please use the right ceph-ansible version for deploying Ceph Nautilus.

You're currently using the master branch which isn't compatible with nautilus.
You should use the stable-4.0 branch.

https://docs.ceph.com/ceph-ansible/master/#releases

**Comment by mafonso** [comment]
I saw this with `ceph version 15.2.1 (9fd2f65f91d9246fae2c841a6222d34d121680ee) octopus (stable)` and the ` stable-5.0` branch.

The issue seems to be that `osdmap` is nested into another `osdmap`

```
ceph -s -f json | jq '.osdmap'
{
  "osdmap": {
    "epoch": 411,
    "num_osds": 4,
    "num_up_osds": 4,
    "num_in_osds": 4,
    "num_remapped_pgs": 0
  }
}
```

so I patched it with 

```
-    - (wait_for_all_osds_up.stdout | from_json)["osdmap"]["num_osds"] | int > 0
-    - (wait_for_all_osds_up.stdout | from_json)["osdmap"]["num_osds"] == (wait_for_all_osds_up.stdout | from_json)["osdmap"]["num_up_osds"]
+    - (wait_for_all_osds_up.stdout | from_json)["osdmap"]["osdmap"]["num_osds"] | int > 0
+    - (wait_for_all_osds_up.stdout | from_json)["osdmap"]["osdmap"]["num_osds"] == (wait_for_all_osds_up.stdout | from_json)["osdmap"]["osdmap"]["num_up_osds"]
```

**Comment by dsavineau** [comment]
> I saw this with ceph version 15.2.1 (9fd2f65f91d9246fae2c841a6222d34d121680ee) octopus (stable) and the stable-5.0 branch.

That could not be true.

The double nested osdmap is present only until nautilus (included). The change has been introduced in octopus and available since 15.1.0 [1]. This has been adated in ceph-ansible via [2] and only present in stable-5.0 and master branches.

So if you see this with stable-5.0 branch then that mean you're not using octopus.

[1] https://github.com/ceph/ceph/commit/f222c7bb1790f4ab3e8e0afd0c7915ed84f0c6c8
[2] https://github.com/ceph/ceph-ansible/commit/a5e359ee806d0229154f644cfe598a46508f502d

**Comment by mafonso** [comment]
> So if you see this with stable-5.0 branch then that mean you're not using octopus.

And I wasn't :) 
I was using an octopus ceph client, but against a server that had not been upgraded yet. 
All was as it supposed to be.
